services/src/main/java/org/keycloak/services/resources/admin/RealmLocalizationResource.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.resources.admin;
+
+import org.jboss.resteasy.plugins.providers.multipart.InputPart;
+import org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataInput;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.cache.CacheRealmProvider;
+import org.keycloak.services.resources.admin.permissions.AdminPermissionEvaluator;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.PATCH;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+
+public class RealmLocalizationResource {
+    private final RealmModel realm;
+    private final AdminPermissionEvaluator auth;
+
+    @Context
+    protected KeycloakSession session;
+
+    public RealmLocalizationResource(RealmModel realm, AdminPermissionEvaluator auth) {
+        this.realm = realm;
+        this.auth = auth;
+    }
+
+    @Path("{locale}/{key}")
+    @POST
+    @Consumes(MediaType.TEXT_PLAIN)
+    public void addRealmLocalizationText(@PathParam("locale") String locale, @PathParam("key") String key,
+            String text) {
+        this.auth.realm().requireManageRealm();
+        try {
+            session.realmLocalizationProvider().saveLocalizationText(realm.getId(), locale, key, text);
+        } catch (ModelDuplicateException e) {
+            throw new BadRequestException(
+                    String.format("Localization text %s for the locale %s and realm %s already exists.",
+                            key, locale, realm.getId()));
+        }
+    }
+
+
+    /**
+     * Import localization from uploaded JSON file
+     */
+    @POST
+    @Path("{locale}")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public void importRealmLocalizationTextsFromFile(@PathParam("locale") String locale, MultipartFormDataInput input)
```

Comment:
In other places, where users upload files, we don't deal with the input stream directly and do not propagate it to downstream layers.

I'm wondering if is not the case of expecting a `Map` already populated from the JSON payload.

Reply:
Changed it :)